### PR TITLE
Ledger: Improve handling of locked Ledger devices

### LINKED ISSUE #1242: Ledger: Handle automatic lock better
When a Ledger Nano S was unlocked and then went into the auto lock, EC does not react to it well. Starting a new EC instance with a Ledger in this state connected results in the following:

```
Traceback (most recent call last):
  File "/home/user/src/electrum/gui/qt/util.py", line 691, in run
    result = task.task()
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 549, in get_client
    client = devmgr.client_for_keystore(self, handler, keystore, force_pair)
  File "/home/user/src/electrum/lib/plugins.py", line 718, in client_for_keystore
    client = self.force_pair_xpub(plugin, handler, info, xpub, derivation, devices)
  File "/home/user/src/electrum/lib/plugins.py", line 748, in force_pair_xpub
    client_xpub = client.get_xpub(derivation, xtype)
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 78, in catch_exception
    return func(self, *args, **kwargs)
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 88, in get_xpub
    self.checkDevice()
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 208, in checkDevice
    raise e
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 204, in checkDevice
    self.perform_hw1_preflight()
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 199, in perform_hw1_preflight
    raise e
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 185, in perform_hw1_preflight
    raise e
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 181, in perform_hw1_preflight
    self.dongleObject.getWalletPublicKey("44'/145'/0'/0/0", showOnScreen=False, cashAddr=True)
  File "/home/user/.local/lib/python3.5/site-packages/btchip/btchip.py", line 124, in getWalletPublicKey
    response = self.dongle.exchange(bytearray(apdu))
  File "/home/user/.local/lib/python3.5/site-packages/btchip/btchipComm.py", line 127, in exchange
    raise BTChipException("Invalid status %04x" % sw, sw)
btchip.btchipException.BTChipException: Exception : Invalid status 6f04
```
After this, when I click the Ledger icon to retry the connection I get this:
```
Traceback (most recent call last):
  File "/home/user/src/electrum/gui/qt/main_window.py", line 82, in onPress
    self.func()
  File "/home/user/src/electrum/plugins/hw_wallet/qt.py", line 230, in show_settings_dialog
    device_id = self.choose_device(window, keystore)
  File "/home/user/src/electrum/plugins/hw_wallet/qt.py", line 223, in choose_device
    info = self.device_manager().select_device(self, keystore.handler, keystore)
  File "/home/user/src/electrum/lib/plugins.py", line 786, in select_device
    infos = self.unpaired_device_infos(handler, plugin, devices)
  File "/home/user/src/electrum/lib/plugins.py", line 775, in unpaired_device_infos
    client = self.create_client(device, handler, plugin)
  File "/home/user/src/electrum/lib/plugins.py", line 651, in create_client
    client = plugin.create_client(device, handler)
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 522, in create_client
    client = self.get_btchip_device(device)
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 515, in get_btchip_device
    dev.open_path(device.path)
  File "hid.pyx", line 72, in hid.device.open_path
OSError: open failed
```
Seems that the device was not properly released and can't be reopened. After restarting EC I then get the following:
```
Traceback (most recent call last):
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 149, in perform_hw1_preflight
    firmwareInfo = self.dongleObject.getFirmwareVersion()
  File "/home/user/.local/lib/python3.5/site-packages/btchip/btchip.py", line 561, in getFirmwareVersion
    response = self.dongle.exchange(bytearray(apdu))
  File "/home/user/.local/lib/python3.5/site-packages/btchip/btchipComm.py", line 127, in exchange
    raise BTChipException("Invalid status %04x" % sw, sw)
btchip.btchipException.BTChipException: Exception : Invalid status 6faa

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/user/src/electrum/gui/qt/util.py", line 691, in run
    result = task.task()
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 549, in get_client
    client = devmgr.client_for_keystore(self, handler, keystore, force_pair)
  File "/home/user/src/electrum/lib/plugins.py", line 718, in client_for_keystore
    client = self.force_pair_xpub(plugin, handler, info, xpub, derivation, devices)
  File "/home/user/src/electrum/lib/plugins.py", line 748, in force_pair_xpub
    client_xpub = client.get_xpub(derivation, xtype)
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 78, in catch_exception
    return func(self, *args, **kwargs)
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 88, in get_xpub
    self.checkDevice()
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 204, in checkDevice
    self.perform_hw1_preflight()
  File "/home/user/src/electrum/plugins/ledger/ledger.py", line 192, in perform_hw1_preflight
    raise Exception("Dongle is temporarily locked - please unplug it and replug it again")
Exception: Dongle is temporarily locked - please unplug it and replug it again
```

### PR DESCRIPTION
**Pull request by EchterAgo**
closes #1242

### CODE REVIEW

**Comment by EchterAgo** [comment]
@cculianu This is ready now

File: plugins/ledger/ledger.py [hunk 1]
```diff
@@ -176,17 +177,8 @@ def perform_hw1_preflight(self):
...
-                except BTChipException as e:
-                    # This call fails on HW1/Nano with a specific error, due to it not supporting cashaddr
-                    if (e.sw != 0x6b00):
-                        raise e
-                self.cashaddrSWSupported = True
-            except TypeError:
-                self.cashaddrSWSupported = False
-
+            gwpkArgSpecs = inspect.getfullargspec(self.dongleObject.getWalletPublicKey)
+            self.cashaddrSWSupported = 'cashAddr' in gwpkArgSpecs.args
```
Comment (cculianu) [inline plugins/ledger/ledger.py]: Yep. I think this is cleaner... as long as you say it's a good test (ie the API doesn't give you a dongleObject that accepts the key if the device does not)...

**Comment by cculianu** [comment]
Cool man.  Looks good and you also did it so fast.  Merging.

**Comment by cculianu** [comment]
Let me know how you want to be credited in the RELEASE-NOTES -- whether it be as EchterAgo or by a real name. 👍

**Comment by EchterAgo** [comment]
Credit me as "Axel Gembe"

I'd still would have liked someone else to test this before merging. I've tested it so far on Windows with Nano S & Nano on Testnet and Mainnet. I'm now going to test it on Linux.

**Comment by cculianu** [comment]
Ok, will do.  Thanks so much for this man.

As for testing: Well I lack a ledger... man if only I had one.... I'll get one soon enough though.

We can always push another commit whenever you like.  If you do find a bug or realize you wanted to do something differently just issue another PR and I'll be resoponsive.  I'm 100% relying on your expertise here without the ability to verify due to a lack of device so -- that at least makes you the final decider, thus making it quick. :)

And.. the way I see it is: you looked at Electrum for inspiration and the code reads decently.  Hopefully it's an improvement over what we had.  

I'll ask in the Slack if anyone is around and can test (are you in the slack? Do you want an invite? It's mostly quiet but it's full of lurkers)...

**Comment by EchterAgo** [comment]
I am on slack already (username is ichundes), but don't have it open all the time due to RAM usage :) Hopefully I'll upgrade my computer soon with lots of RAM.

I am quite confident in the changes I made, but additional testing would still be appreciated. Does EC do prereleases for users to test?

**Comment by cculianu** [comment]
Yeah slack leaks memory like it's old Firefox circa 2010.  :P  So does its derivative, Discord.   Whereas Telegram is light and fast.

> Does EC do prereleases for users to test?

Sometimes if it's a huge change like cashshuffle.  I could throw a pre-release up now on the releases page if you wanna shill it somewhere... should I do that?

**Comment by EchterAgo** [comment]
It will probably be fine without.

**Comment by cculianu** [comment]
Yeah no worries -- if there's a bug we can fix it quickly.

**Pull request merged**